File: leetcode/522.最长特殊序列-ii.cpp
#include <iostream>
#include <vector>
using namespace std;
/*
 * @lc app=leetcode.cn id=522 lang=cpp
 *
 * [522] 最长特殊序列 II
 */

// @lc code=start
class Solution
{
public:
    int findLUSlength(vector<string> &strs)
    {
        int res = -1;
        int n = strs.size();
        for (size_t i = 0; i < n; i++)
        {
            bool check = true;
            for (size_t j = 0; j < n; j++)
            {
                if (i != j && is_sub(strs[j], strs[i]))
                {
                    check = false;
                    break;
                }
            }
            if (check)
                res = max(res, static_cast<int>(strs[i].size()));
        }
        return res;
    }
    bool is_sub(string &word1, string &word2)
    {
        int i = 0, j = 0;
        int n = word1.size();
        int m = word2.size();
        while (i < n && j < m)
        {
            if (word1[i] == word2[j])
                j++;
            i++;
        }
        return j == m;
    }
};
// @lc code=end
int main()
{
    Solution s = Solution();
    string strs[] = {"aba", "cdc", "eae"};
    vector<string> vect(strs, strs + 3);
    cout << s.findLUSlength(vect) << endl;
}

File: leetcode/522.最长特殊序列-ii.py
#
# @lc app=leetcode.cn id=522 lang=python3
#
# [522] 最长特殊序列 II
#
from typing import List


# @lc code=start
class Solution:
    def findLUSlength(self, strs: List[str]) -> int:
        def is_sub(word1: str, word2: str) -> bool:
            """ return id word2 is sub str of word1"""
            start1 = start2 = 0
            end1 = len(word1)
            end2 = len(word2)
            while start1 < end1 and start2 < end2:
                if word1[start1] == word2[start2]:
                    start2 += 1
                start1 += 1
            return start2 == end2

        res = -1

        for i in range(len(strs)):
            check = True
            for j in range(len(strs)):
                if i != j and is_sub(strs[j], strs[i]):
                    check = False
                    break
            if check:
                res = max(res, len(strs[i]))

        return res


# @lc code=end
print(Solution().findLUSlength(["aba", "cdc", "eae"]))
print(Solution().findLUSlength(["aaa", "aaa", "aa"]))
